<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Detecting CNV from SNPs: a quick guide to rCNV • rCNV</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Detecting CNV from SNPs: a quick guide to rCNV">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rCNV</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/rCNV.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/piyalkarum/rCNV/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Detecting CNV from SNPs: a quick guide to rCNV</h1>
                        <h4 data-toc-skip class="author">Piyal
Karunarathne</h4>
            
            <h4 data-toc-skip class="date">Last updated: 16 September,
2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/piyalkarum/rCNV/blob/HEAD/vignettes/rCNV.Rmd" class="external-link"><code>vignettes/rCNV.Rmd</code></a></small>
      <div class="d-none name"><code>rCNV.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="rcnv-an-r-package-for-detecting-copy-number-variants-from-snps-data">
<strong>rCNV: An R package for detecting copy number variants from
SNPs data</strong><a class="anchor" aria-label="anchor" href="#rcnv-an-r-package-for-detecting-copy-number-variants-from-snps-data"></a>
</h2>
<p><span style="color: dodgerblue;">Piyal Karunarathne, Qiujie Zhou, and
Pascal Milesi</span></p>
<p>This tutorial provides a complete guide to detecting CNVs from SNPs
data using rCNV package. This includes importing raw VCF files,
filtering them, normalization of read depth, categorization of SNPs into
“deviants” and “non-deviants”, and from which SNPs located in putative
CNV regions are filtered. A brief workflow is presented in Fig 1.
below.</p>
<div class="float">
<img src="workflow2.png" width="400" alt="Fig. 1. Workflow of detecting CNVs with rCNV"><div class="figcaption">Fig. 1. Workflow of detecting CNVs with
rCNV</div>
</div>
<p>All the data used in this tutorial is available either withing the
package or from the GitHub repository at <a href="https://github.com/piyalkarum/rCNV/tree/master/testDATA" class="external-link uri">https://github.com/piyalkarum/rCNV/tree/master/testDATA</a>.</p>
<p>We provide here a complete workflow of detecting CNVs from SNPs,
starting with raw (unfiltered) VCF as in the number 1 of the workflow
chart. However, if one believes that they have filtered VCFs for the
parameters we highlight (see Fig.1 and sub-sections 1.2 and 1.3), they
can start from the number 2 in the workflow shown above.<br><em>NOTE</em>: In the latest iteration of the rCNV package, the
functions <code><a href="../reference/sig.hets.html">sig.hets()</a></code> and <code><a href="../reference/dupGet.html">dupGet()</a></code> use use a
sample based inbreeding coefficient <em>(Fis)</em> to improve the
accuracy of the deviant detection. This <em>Fis</em> is calculated using
the <code><a href="../reference/h.zygosity.html">h.zygosity()</a></code> function; see below for more
information.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Start by installing the package if you haven't already done so.  </span></span>
<span><span class="co"># To install the CRAN version</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rCNV"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or if you would like to install the development version of the package</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span> </span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"piyalkarum/rCNV"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="pre-detection">1. PRE-DETECTION<a class="anchor" aria-label="anchor" href="#pre-detection"></a>
</h3>
<div class="section level4">
<h4 id="importing-data">1.1 Importing data<a class="anchor" aria-label="anchor" href="#importing-data"></a>
</h4>
<div class="line-block">    Raw VCF files can be imported to rCNV R
environment using the built-in function <code><a href="../reference/readVCF.html">readVCF()</a></code>. The
imported vcf is stored as a data frame within a list.</div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://piyalkarum.github.io/rCNV/" class="external-link">rCNV</a></span><span class="op">)</span></span>
<span><span class="va">vcf.file.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"rCNV"</span><span class="op">)</span>, <span class="st">"/example.raw.vcf.gz"</span><span class="op">)</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readVCF.html">readVCF</a></span><span class="op">(</span><span class="va">vcf.file.path</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This imports an example VCF file included in the package</span></span>
<span><span class="co"># The data is a partial dataset of Norway spruce data from Chen et al. 2019.</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#print the first 10 rows and columns of the imported VCF</span></span>
<span><span class="va">vcf</span><span class="op">$</span><span class="va">vcf</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;    #CHROM   POS     ID    REF    ALT      QUAL FILTER                 FORMAT</span></span>
<span><span class="co">#&gt;    &lt;char&gt; &lt;int&gt; &lt;char&gt; &lt;char&gt; &lt;char&gt;     &lt;num&gt; &lt;char&gt;                 &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:  MA_10 24959      .      A      C   4338.29   PASS GT:AD:DP:GQ:PGT:PID:PL</span></span>
<span><span class="co">#&gt; 2:  MA_10 26811      .      C      T  33278.10   PASS         GT:AD:DP:GQ:PL</span></span>
<span><span class="co">#&gt; 3:  MA_10 26860      .      A      G 273719.00   PASS GT:AD:DP:GQ:PGT:PID:PL</span></span>
<span><span class="co">#&gt; 4:  MA_10 26874      .      T      C 374526.00   PASS GT:AD:DP:GQ:PGT:PID:PL</span></span>
<span><span class="co">#&gt; 5:  MA_31 21911      .      C      G   2069.06   PASS         GT:AD:DP:GQ:PL</span></span>
<span><span class="co">#&gt; 6:  MA_31 40129      .      A      T   6549.58   PASS         GT:AD:DP:GQ:PL</span></span>
<span><span class="co">#&gt;                              AT_PA_06_12</span></span>
<span><span class="co">#&gt;                                   &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:             0/0:7,0:7:21:.:.:0,21,269</span></span>
<span><span class="co">#&gt; 2:               0/0:18,0:18:22:0,22,576</span></span>
<span><span class="co">#&gt; 3: 0/1:9,5:14:99:0|1:26860_A_G:183,0,775</span></span>
<span><span class="co">#&gt; 4:             0/0:15,0:15:3:.:.:0,3,407</span></span>
<span><span class="co">#&gt; 5:               0/0:11,0:11:33:0,33,319</span></span>
<span><span class="co">#&gt; 6:               0/0:11,0:11:33:0,33,347</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="filtering-vcf-for-missing-data">1.2 Filtering VCF for missing data<a class="anchor" aria-label="anchor" href="#filtering-vcf-for-missing-data"></a>
</h4>
<div class="line-block">      The raw VCF files often contain missing
data for both samples and SNPs. If retained, these samples and SNPs can
introduce false positives and negatives to the analysis. Therefore, they
must be removed based on a missing percentage. The function
<code><a href="../reference/get.miss.html">get.miss()</a></code> determines the missing percentages in samples
and SNPs and graphically visualize the missing ranges with 5% quantile
values. We recommend removing samples and SNPs with missing percentages
higher than 50%.</div>
<p>The following example demonstrates such filtering on Parrot Fish data
from <em>Stockwell et al. (2016)</em>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#download the parrotfish data (Stockwell et al. 2016) file from the following link to your working directory if you do not have the github folder</span></span>
<span><span class="co">#https://github.com/piyalkarum/rCNV/blob/master/testDATA/Parrotfish_sequenceReads.vcf.gz</span></span>
<span><span class="co">#and specify the path fl= to your file.</span></span>
<span><span class="va">fl</span><span class="op">&lt;-</span><span class="st">"../testDATA/Parrotfish_sequenceReads.vcf.gz"</span></span>
<span><span class="va">parrot</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/readVCF.html">readVCF</a></span><span class="op">(</span><span class="va">fl</span>,verbose<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">mss</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/get.miss.html">get.miss</a></span><span class="op">(</span><span class="va">parrot</span>,verbose <span class="op">=</span> <span class="cn">F</span>, plot<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>This generates a list with missing percentage tables for SNPs and
samples and a plot similar to Fig 1.2 where missing percentages are
plotted. Users can also manually plot these percentages on their
discretion. Further, if desired one type of missing percentage (i.e. per
sample or per SNP) can be calculated separately by specifying the
argument <code>type=</code> in the <code><a href="../reference/get.miss.html">get.miss()</a></code> function. See
the help page of <code><a href="../reference/get.miss.html">get.miss()</a></code> for more information.</p>
<div class="float">
<img src="get.miss.jpeg" style="width:80.0%" alt="Fig.1.2. Missing status of the VCF"><div class="figcaption">Fig.1.2. Missing status of the VCF</div>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mss</span><span class="op">$</span><span class="va">perSample</span><span class="op">)</span></span>
<span><span class="co">#&gt;   indiv n_miss     f_miss</span></span>
<span><span class="co">#&gt; 1  DP_1   1089 0.05726455</span></span>
<span><span class="co">#&gt; 2 DP_10   1086 0.05710680</span></span>
<span><span class="co">#&gt; 3 DP_11   1365 0.07177788</span></span>
<span><span class="co">#&gt; 4 DP_12   1091 0.05736972</span></span>
<span><span class="co">#&gt; 5 DP_13   1069 0.05621286</span></span>
<span><span class="co">#&gt; 6 DP_14    387 0.02035021</span></span></code></pre></div>
<p>The plot shows that there are several samples with missing data for
more than 50% of the SNPs. It is recommended to remove these before
proceeding with further analysis. Removing them from the imported vcf
file is just a matter of obtaining the columns (samples) for which
missing percentage if &gt;50%.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">$</span><span class="va">vcf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19017   105</span></span>
<span><span class="va">sam</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">mss</span><span class="op">$</span><span class="va">perSample</span><span class="op">$</span><span class="va">f_miss</span><span class="op">&gt;</span><span class="fl">0.5</span><span class="op">)</span><span class="op">+</span><span class="fl">9</span> <span class="co">#here we determine the sample columns with f_miss&gt;0.5 and add 9 to get the correct column number because there are 9 more columns before the sample columns start.</span></span>
<span><span class="va">parrot</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">$</span><span class="va">vcf</span><span class="op">)</span><span class="op">[</span>,<span class="op">-</span><span class="va">sam</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19017   100</span></span></code></pre></div>
<p>We have removed 5 samples that were undesired for the analysis due to
their missing data. Undesired SNPs can also be removed the same way;
Note: now they are in rows.</p>
</div>
<div class="section level4">
<h4 id="assessing-the-relatedness-among-samples-and-heterozygosity-within">1.3 Assessing the relatedness among samples and heterozygosity
within<a class="anchor" aria-label="anchor" href="#assessing-the-relatedness-among-samples-and-heterozygosity-within"></a>
</h4>
<p>The function <code><a href="../reference/relatedness.html">relatedness()</a></code> in the rCNV package will
generate the relatedness index between pairs of samples according to
<em>Yang et al. (2010)</em>. The output is a data frame with pairs of
samples and the relatedness score <em>Ajk</em>. Please read <em>Yang et
al. (2010)</em> for more information on the calculation of
<em>Ajk</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rels</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/relatedness.html">relatedness</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rels</span><span class="op">)</span></span></code></pre></div>
<p>From the relatedness plots, we can see that the relatedness in Parrot
fish samples is low. We recommend removing one sample of the pairs with
outlier values above 0.9.</p>
<p><img src="relatedness.png" style="width:40.0%"><img src="h.zygosity.png" style="width:40.0%"></p>
<p>Figure 1.3. a. Relatedness among samples, b. Heterozygosity in
Parrotfish populations</p>
<p>Heterozygosity within populations can be calculated using
<code><a href="../reference/h.zygosity.html">h.zygosity()</a></code> function.</p>
<div class="line-block">  * NOTE: A list of populations corresponding to
the samples must be provided in order to plot the heterozygosity among
populations.</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span> <span class="co">#population codes (the first two letters of the sample names)</span></span>
<span><span class="va">hz</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/h.zygosity.html">h.zygosity</a></span><span class="op">(</span><span class="va">parrot</span>,pops<span class="op">=</span><span class="va">pops</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hz</span><span class="op">)</span></span></code></pre></div>
<p>In figure 1.3.b, we can see that most of the individuals has a mean
heterozygosity F<sub>IS</sub> close to 0.19, except for a few outliers.
However, outliers with an F<sub>IS</sub> value below -0.2 is highly
likely DNA contamination. Therefore, it is recommended to remove such
samples.</p>
</div>
</div>
<div class="section level3">
<h3 id="generating-allele-depth-tables-and-normalized-depth-values">1.4 Generating allele depth tables and normalized depth values<a class="anchor" aria-label="anchor" href="#generating-allele-depth-tables-and-normalized-depth-values"></a>
</h3>
<p>Allele read depth table is the main data that will be used for the
rest of the analysis. Therefore, having them generated and stored for
later use will immensely save a lot of time. The function
<code><a href="../reference/hetTgen.html">hetTgen()</a></code> is especially dedicated for this purpose. Apart
from generating filtered allele depth table for separate alleles
(<code>info.type="AD"</code>), the function can also generate total
allele depth (<code>info.type="AD-tot"</code>), genotype table
(<code>info.type="GT"</code>), etc.. See the help page of
<code><a href="../reference/hetTgen.html">hetTgen()</a></code> for more formats.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ad.tab</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/hetTgen.html">hetTgen</a></span><span class="op">(</span><span class="va">parrot</span>,info.type<span class="op">=</span><span class="st">"AD"</span><span class="op">)</span></span>
<span><span class="va">ad.tab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
Table 1.4.0 Allele depth table (generated from a larger data set of
Stockwell et al. 2016 Parrot fish data)
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
CHROM
</th>
<th style="text-align:right;">
POS
</th>
<th style="text-align:right;">
ID
</th>
<th style="text-align:left;">
ALT
</th>
<th style="text-align:left;">
KKOGR07_0005_combo
</th>
<th style="text-align:left;">
KKOGR07_0006
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
7452
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
705339
</td>
<td style="text-align:right;">
67808
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
45,0
</td>
<td style="text-align:left;">
34,0
</td>
</tr>
<tr>
<td style="text-align:left;">
18079
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
769896
</td>
<td style="text-align:right;">
80692
</td>
<td style="text-align:left;">
T
</td>
<td style="text-align:left;">
16,0
</td>
<td style="text-align:left;">
15,0
</td>
</tr>
<tr>
<td style="text-align:left;">
9793
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
926883
</td>
<td style="text-align:right;">
109573
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
56,0
</td>
<td style="text-align:left;">
42,0
</td>
</tr>
<tr>
<td style="text-align:left;">
11755
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
172395
</td>
<td style="text-align:right;">
16363
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
61,67
</td>
<td style="text-align:left;">
40,34
</td>
</tr>
<tr>
<td style="text-align:left;">
10122
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
18102
</td>
<td style="text-align:right;">
1703
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
75,0
</td>
<td style="text-align:left;">
51,0
</td>
</tr>
<tr>
<td style="text-align:left;">
816
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
76725
</td>
<td style="text-align:right;">
6991
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
66,41
</td>
<td style="text-align:left;">
20,20
</td>
</tr>
</tbody>
</table>
<p>As an additional step, rCNV framework corrects allele depth values
for genotype misclassifications (where genotype and the depth values do
not match) and odd number anomalies (where total depth of a loci is
odd).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#the genotype table is needed for correcting the genotype mismatches</span></span>
<span><span class="va">gt</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/hetTgen.html">hetTgen</a></span><span class="op">(</span><span class="va">parrot</span>,info.type <span class="op">=</span> <span class="st">"GT"</span><span class="op">)</span></span>
<span><span class="va">ad.tab</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/ad.correct.html">ad.correct</a></span><span class="op">(</span><span class="va">ad.tab</span>,gt.table <span class="op">=</span> <span class="va">gt</span><span class="op">)</span></span></code></pre></div>
<p>It is recommended to normalize the allele depth values across samples
in order to avoid the depth variation arise from sequencing
heterogeneity where some samples tend to have higher depth values
compared to others due to different origin of sequencing efforts. We
provide several methods for normalizing data.<br>
| TMM = trimmed mean of M values | MedR = median ratio normalization |
QN = quantile based normalization | pca = PCA based normalization</p>
<p>*Please read the function help for more information on the different
methods</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#normalize depth table with cpm.normal()</span></span>
<span><span class="va">ad.nor</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/cpm.normal.html">cpm.normal</a></span><span class="op">(</span><span class="va">ad.tab</span>,method<span class="op">=</span><span class="st">"MedR"</span><span class="op">)</span></span>
<span><span class="va">ad.nor</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<p>Table 1.4.1 Normalized allele depth table</p>
<pre><code><span><span class="co">#&gt; OUTLIERS DETECTED</span></span>
<span><span class="co">#&gt; Consider removing the samples:</span></span>
<span><span class="co">#&gt; KKOGR07_0030 KKOGR07_0043_combo KKOKT10_0069</span></span></code></pre>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
CHROM
</th>
<th style="text-align:right;">
POS
</th>
<th style="text-align:right;">
ID
</th>
<th style="text-align:left;">
ALT
</th>
<th style="text-align:left;">
KKOGR07_0005_combo
</th>
<th style="text-align:left;">
KKOGR07_0006
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
7452
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
705339
</td>
<td style="text-align:right;">
67808
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
29,0
</td>
<td style="text-align:left;">
34,0
</td>
</tr>
<tr>
<td style="text-align:left;">
18079
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
769896
</td>
<td style="text-align:right;">
80692
</td>
<td style="text-align:left;">
T
</td>
<td style="text-align:left;">
10,0
</td>
<td style="text-align:left;">
15,0
</td>
</tr>
<tr>
<td style="text-align:left;">
9793
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
926883
</td>
<td style="text-align:right;">
109573
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
36,0
</td>
<td style="text-align:left;">
42,0
</td>
</tr>
<tr>
<td style="text-align:left;">
11755
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
172395
</td>
<td style="text-align:right;">
16363
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
39,43
</td>
<td style="text-align:left;">
40,34
</td>
</tr>
<tr>
<td style="text-align:left;">
10122
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
18102
</td>
<td style="text-align:right;">
1703
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
48,0
</td>
<td style="text-align:left;">
51,0
</td>
</tr>
<tr>
<td style="text-align:left;">
816
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
76725
</td>
<td style="text-align:right;">
6991
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
42,26
</td>
<td style="text-align:left;">
20,20
</td>
</tr>
</tbody>
</table>
<p>Compare the allele depth values between tables 1.4.0 and 1.4.1.</p>
</div>
<div class="section level3">
<h3 id="detection">2. DETECTION<a class="anchor" aria-label="anchor" href="#detection"></a>
</h3>
<div class="section level4">
<h4 id="generating-allele-information-table">2.1 Generating allele information table<a class="anchor" aria-label="anchor" href="#generating-allele-information-table"></a>
</h4>
<p>In this step, all the statistics necessary for detecting duplicates
are calculated. They include the following: a. Allele ratios (across all
samples), b. Proportions of homo-/hetero-zygotes per SNP, c. Depth
ratios (alternative/sum), d. Z-score, e. Chi-square significance, and
f. excess of heterozygosity</p>
<p>Alternative allele ratio is calculated across all the samples from
depth values; Proportion of homo/hetero-zygotes are also calculated
across all the samples per SNP; Depth ratio is calculated per individual
per SNP by dividing the alternative allele depth value by the total
depth value of both alleles. This is calculated for both heterozygotes
and homozygotes; the Z-score per SNP is calculated according to the
following equation:</p>
<p><img src="Z2.1.svg"></p>
<p><img src="Z2.2.svg"></p>
<p>Where <em>N<sub>i</sub></em> is the total depth for heterozygote
<em>i</em> at SNP <em>x</em>, <em>N<sub>Ai</sub></em> is the alternative
allele read depth for heterozygote <em>i</em> at SNP <em>x</em>,
<em>p</em> is the probability of sampling allele <em>A</em> in SNP
<em>x</em> - for unbiased sequencing, this is <em>0.5</em>. The
<code><a href="../reference/allele.info.html">allele.info()</a></code> function calculates this for both
<em>0.5</em> and biased probability using the ratio between reference
and alternative alleles.</p>
<p>The Chi-squared values per SNP per sample were calculated using the
following equation:</p>
<p><img src="chi2.1.svg"></p>
<p><img src="chi2.2.svg"></p>
<p>Where <em>N<sub>i</sub></em> is the total depth for heterozygote
<em>i</em> at SNP <em>x</em>, <em>N<sub>Ai</sub></em> is the alternative
allele read-depth for heterozygote <em>i</em> at SNP <em>x</em>,
<em>p</em> is the probability of sampling allele <em>A</em> at SNP
<em>x</em> in heterozygotes - in unbiased sequencing, this is
<em>0.5</em>, <em>n</em> is the number of heterozygotes at SNP
<em>x</em>.</p>
<p>In addition to the above a-e parameters several other important
statistics per sample per SNP is generated in the
<code><a href="../reference/allele.info.html">allele.info()</a></code> output.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># in the new version of the package allele.info() and dupGet() functions require Fis value for a more accurate detection of the deviants</span></span>
<span><span class="co"># this is obtained by using the h.zygosity() function as below</span></span>
<span><span class="va">hz</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/h.zygosity.html">h.zygosity</a></span><span class="op">(</span><span class="va">parrot</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co">#parrot is the filtered vcf file</span></span>
<span><span class="va">Fis</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">hz</span><span class="op">$</span><span class="va">Fis</span>,na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">A.info</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/allele.info.html">allele.info</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">ADtable</span>,x.norm <span class="op">=</span> <span class="va">ADnorm</span>, plot.allele.cov <span class="op">=</span> <span class="cn">TRUE</span>,Fis<span class="op">=</span><span class="va">Fis</span><span class="op">)</span></span>
<span><span class="co"># X is the corrected non-normalized allele depth table and x.norm is the normalized allele depth table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">A.info</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;       CHROM    POS     ID NHet    propHet  medRatio NHomRef NHomAlt propHomAlt</span></span>
<span><span class="co">#&gt; 7452     un 705339  67808    9 0.08653846 0.5000000      95       0 0.00000000</span></span>
<span><span class="co">#&gt; 18079    un 769896  80692   41 0.40594059 0.5000000      48      12 0.11881188</span></span>
<span><span class="co">#&gt; 9793     un 926883 109573   54 0.51923077 0.5096539      34      16 0.15384615</span></span>
<span><span class="co">#&gt; 10122    un  18102   1703    8 0.07619048 0.5174242      97       0 0.00000000</span></span>
<span><span class="co">#&gt; 816      un  76725   6991   49 0.47115385 0.4897959      35      20 0.19230769</span></span>
<span><span class="co">#&gt; 17482    un 713297  72437   46 0.43809524 0.5217593      52       7 0.06666667</span></span>
<span><span class="co">#&gt;       Nsamp      pAll      pHet       fis     z.het      z.05     z.all</span></span>
<span><span class="co">#&gt; 7452    104 0.5514787 0.5154153 0.9899497 0.9816955 0.5373322 0.1694981</span></span>
<span><span class="co">#&gt; 18079   101 0.5079704 0.4822855 0.9907917 0.9630120 0.3160515 0.1518454</span></span>
<span><span class="co">#&gt; 9793    104 0.5043221 0.4939181 0.9897064 0.6531659 0.8933630 0.5832463</span></span>
<span><span class="co">#&gt; 10122   105 0.4610830 0.4917005 0.9900990 0.7954672 0.5758174 0.3945965</span></span>
<span><span class="co">#&gt; 816     104 0.5007013 0.5003297 0.9907469 0.6343455 0.6562743 0.6100144</span></span>
<span><span class="co">#&gt; 17482   105 0.5085058 0.5166027 0.9897778 0.9822687 0.1245431 0.4469195</span></span>
<span><span class="co">#&gt;         chi.het    chi.05   chi.all   z.het.sum    z.05.sum z.all.sum</span></span>
<span><span class="co">#&gt; 7452  0.9175753 0.9061279 0.8927433 -0.06882976  -1.8505568  4.121455</span></span>
<span><span class="co">#&gt; 18079 0.9999231 0.9999047 0.9998801  0.29693961   6.4198038  9.175957</span></span>
<span><span class="co">#&gt; 9793  0.9999640 0.9999651 0.9999629 -3.30217028   0.9850626  4.031731</span></span>
<span><span class="co">#&gt; 10122 0.7615202 0.7593583 0.7221300  0.73317868   1.5825180 -2.407872</span></span>
<span><span class="co">#&gt; 816   0.9995718 0.9995749 0.9995681  3.32933637   3.1154399  3.570371</span></span>
<span><span class="co">#&gt; 17482 0.9996000 0.9992832 0.9995143 -0.15073519 -10.4175272 -5.158373</span></span>
<span><span class="co">#&gt;       chi.het.sum chi.05.sum chi.all.sum   eH.pval   eH.delta        cv</span></span>
<span><span class="co">#&gt; 7452     3.251816   3.409112    3.581875 0.6446423  0.3894231 1.0112356</span></span>
<span><span class="co">#&gt; 18079   14.588811  14.828298   15.091843 0.4819957 -3.0841584 0.9649802</span></span>
<span><span class="co">#&gt; 9793    21.505630  21.466620   21.545469 0.4719765  3.5576923 0.7752089</span></span>
<span><span class="co">#&gt; 10122    4.157058   4.175480    4.488264 0.6848750  0.3047619 1.0355910</span></span>
<span><span class="co">#&gt; 816     21.872400  21.860947   21.885888 0.7008333 -1.9182692 0.7840350</span></span>
<span><span class="co">#&gt; 17482   19.798998  20.702798   20.092246 0.4523861  3.1428571 0.8057838</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="deviants-detection">2.2 Deviants detection<a class="anchor" aria-label="anchor" href="#deviants-detection"></a>
</h4>
<p>In this step, the allele information table generated in the previous
step is used to detect deviant SNPs based on different models
(statistics), which then is used in the next step with the appropriate
test statistic to filter putative CNVs. The rCNV function
<code><a href="../reference/dupGet.html">dupGet()</a></code> delivers easy to pick model selection to detect
deviant SNPs with a plotting option to visualize the detection.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run this code for a demonstration of the detection</span></span>
<span><span class="co"># Fis is calculated as before using h.zygosity()</span></span>
<span><span class="va">dvs</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/dupGet.html">dupGet</a></span><span class="op">(</span><span class="va">alleleINF</span>,test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"z.05"</span>,<span class="st">"chi.05"</span><span class="op">)</span>,Fis<span class="op">=</span><span class="fl">0.11</span><span class="op">)</span></span>
<span><span class="co"># z score and chi-square values given p=0.05 is used here because the data is RADseq generated and probe-biase is neglegible</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dvs</span><span class="op">)</span></span></code></pre></div>
<p>The detection of deviant SNPs is based on the
<strong>heterozygosity</strong> within populations (or among all
samples) and <strong>allelic ratios</strong> within such heterozygotes.
As such, <code><a href="../reference/dupGet.html">dupGet()</a></code> function implements three methods to flag
the deviant alleles; They are a) Excess of heterozygotes, b)
Heterozygous SNPs with depth values that do not follow normal Z-score
distribution, and c) Heterozygous SNPs with depth values that do not
follow normal chi-square distribution.</p>
<div class="line-block">  The proportion of heterozygotes in a
population according to HWE is 2pq where as homozygotes is
q<sup>2</sup>. In deviant SNPs heterozygosity is expected to be higher
indicating excess of heterozygosity and thus flagged as
deviant/putative-CNVs.</div>
<div class="line-block">  In non-duplicated alleles (non-CNVs), median
allele ratio must be ~0.5, however, in deviants, it varies between 1:3
and 3:1. These SNP deviants are also flagged as putative CNVs. To
determine the deviation, we use the Z-score values and Chi-square
significance calculated in the previous step.</div>
<div class="line-block">  **NOTE: users can select the expected allele
frequency (p: ) of the alternative allele based on the sequencing method
and the probe bias therein. See our publication <a href="https://doi.org/10.1101/2022.10.14.512217" class="external-link uri">https://doi.org/10.1101/2022.10.14.512217</a> for more
details on probe bias.</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deviants</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/dupGet.html">dupGet</a></span><span class="op">(</span><span class="va">alleleINF</span>,Fis<span class="op">=</span><span class="va">Fis</span>,test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"z.all"</span>,<span class="st">"chi.all"</span><span class="op">)</span>,plot<span class="op">=</span><span class="cn">TRUE</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deviants</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;       CHROM    POS     ID NHet    propHet  medRatio NHomRef NHomAlt propHomAlt</span></span>
<span><span class="co">#&gt; 7452     un 705339  67808    9 0.08653846 0.5000000      95       0 0.00000000</span></span>
<span><span class="co">#&gt; 18079    un 769896  80692   41 0.40594059 0.5000000      48      12 0.11881188</span></span>
<span><span class="co">#&gt; 9793     un 926883 109573   54 0.51923077 0.5096539      34      16 0.15384615</span></span>
<span><span class="co">#&gt; 10122    un  18102   1703    8 0.07619048 0.5174242      97       0 0.00000000</span></span>
<span><span class="co">#&gt; 816      un  76725   6991   49 0.47115385 0.4897959      35      20 0.19230769</span></span>
<span><span class="co">#&gt; 17482    un 713297  72437   46 0.43809524 0.5217593      52       7 0.06666667</span></span>
<span><span class="co">#&gt;       Nsamp   eH.pval   eH.delta    dup.stat</span></span>
<span><span class="co">#&gt; 7452    104 0.6446423  0.3894231 non-deviant</span></span>
<span><span class="co">#&gt; 18079   101 0.4819957 -3.0841584 non-deviant</span></span>
<span><span class="co">#&gt; 9793    104 0.4719765  3.5576923 non-deviant</span></span>
<span><span class="co">#&gt; 10122   105 0.6848750  0.3047619 non-deviant</span></span>
<span><span class="co">#&gt; 816     104 0.7008333 -1.9182692 non-deviant</span></span>
<span><span class="co">#&gt; 17482   105 0.4523861  3.1428571 non-deviant</span></span></code></pre>
<p>The function also plots the deviants with Allele median ratio Vs
Proportion of Heterozygotes to visualize and validate the detection by
the user. Users can also plot the detection separately with the output
of <code><a href="../reference/dupGet.html">dupGet()</a></code> using <code><a href="../reference/dup.plot.html">dup.plot()</a></code> function.</p>
<div class="float">
<img src="dev.plot2.png" width="400" alt="Fig 2.1 Median allele ratio Vs Prop. of Heterozygotes with deviant alleles"><div class="figcaption">Fig 2.1 Median allele ratio Vs Prop. of
Heterozygotes with deviant alleles</div>
</div>
</div>
<div class="section level4">
<h4 id="filtering-putative-cnvs">2.3 Filtering putative CNVs<a class="anchor" aria-label="anchor" href="#filtering-putative-cnvs"></a>
</h4>
<p>SNPs that are located in putative CNV regions are filtered using two
methods; 1) intersection: using a combination of at least two statistics
used in deviant detection. i.e., excess of heterozygotes, Z-score
distribution and Chi-square distribution, 2) K-means: using an
unsupervised clustering based on Z-score, chi-square, excess of
heterozygotes, and coefficient of variation from read-depth dispersion.
The significant SNPs from either of the <code>filtering</code> are
flagged as putative duplicates. The users can pick a range of optimal
statistics
(e.g. <code>z.all</code>,<code>chi.all</code>,<code>z.05</code>,<code>chi.05</code>,
etc.) depending on the nature (also sequencing technology) of the
underlying data. The function <code><a href="../reference/cnv.html">cnv()</a></code> is dedicated for this
purpose.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CV</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/cnv.html">cnv</a></span><span class="op">(</span><span class="va">alleleINF</span>, test<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"z.05"</span>,<span class="st">"chi.05"</span><span class="op">)</span>, filter <span class="op">=</span> <span class="st">"kmeans"</span><span class="op">)</span></span>
<span><span class="co"># see the difference between deviants and duplicates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">deviants</span><span class="op">$</span><span class="va">dup.stat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">CV</span><span class="op">$</span><span class="va">dup.stat</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="dup.plot.parrotfish.png" width="400" alt="Fig 2.1 Median allele ratio Vs Prop. of Heterozygotes with putative duplicates"><div class="figcaption">Fig 2.1 Median allele ratio Vs Prop. of
Heterozygotes with putative duplicates</div>
</div>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     deviant non-deviant </span></span>
<span><span class="co">#&gt;         335        2522</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     cnv non-cnv </span></span>
<span><span class="co">#&gt;     388    2469</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="post-detection">3. POST-DETECTION<a class="anchor" aria-label="anchor" href="#post-detection"></a>
</h3>
<div class="section level4">
<h4 id="duplicate-list-and-validation">3.1 Duplicate list and validation<a class="anchor" aria-label="anchor" href="#duplicate-list-and-validation"></a>
</h4>
<p>We test the validity of detected duplicates in two ways: 1. direct
detection through a sliding window and 2. variant fixation index</p>
<ul>
<li><p>The sliding window method assess the duplication along
chromosomes/scaffolds on a given size (e.g., 10,000 bp) sliding window.
This step assumes that ideally, if a locus is truly located in a
multi-copy region, close-by SNPs should also be classified as deviants
while deviation caused by sequencing error should be more scattered
along the genome. The putative duplicates that do not follow this
assumption will be flagged as low-confident and can be removed if
desired. The function <code>dup.validate</code> in the rCNV package is
dedicated to detecting regions enriched for deviant SNPs within a
sliding window along a chromosome, scaffold, or a sequence of any given
length.<br>
** This method is still under beta-testing **</p></li>
<li><p>Variant fixation index
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mrow><mi>S</mi><mi>T</mi></mrow></msub><annotation encoding="application/x-tex">V_{ST}</annotation></semantics></math>
: Redon et al. 2006) is analogous to population fixation index
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>F</mi><mrow><mi>S</mi><mi>T</mi></mrow></msub><annotation encoding="application/x-tex">F_{ST}</annotation></semantics></math>)
and calculated as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>V</mi><mrow><mi>S</mi><mi>T</mi></mrow></msub><mo>=</mo><msub><mi>V</mi><mi>S</mi></msub><mi>/</mi><msub><mi>V</mi><mi>T</mi></msub></mrow><annotation encoding="application/x-tex">V_{ST} = V_S/V_T</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>T</mi></msub><annotation encoding="application/x-tex">V_T</annotation></semantics></math>
is the variance of normalized read depths among all individuals from the
two populations and VS is the average of the variance within each
population, weighed for population size; and used to identify distinct
CNV groups between populations (See Dennis et al., 2017; Weir &amp;
Cockerham, 1984) The function <code>vst</code> can be used to calculate
the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mrow><mi>S</mi><mi>T</mi></mrow></msub><annotation encoding="application/x-tex">V_{ST}</annotation></semantics></math>
values per population and plot genetic distance in a network
qgraph.</p></li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the dataset used here is from Dorrant et al. 2020 of American lobster populations in Gulf of St. Lawrence, Canada</span></span>
<span><span class="va">vdata</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../testDATA/lobster_data4Vst.rds"</span><span class="op">)</span></span>
<span><span class="co"># get the total depth values of the duplicated snps</span></span>
<span><span class="co"># You can either use the following line to get the total depth of the samples or use the "tot" option of the hetTgen function to generate the total depth table</span></span>
<span><span class="va">dd</span><span class="op">&lt;-</span><span class="fu">rCNV</span><span class="fu">:::</span><span class="fu">apply_pb</span><span class="op">(</span><span class="va">vdata</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">]</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>,<span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x</span>,<span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,<span class="st">","</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">pops</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dd</span><span class="op">)</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">dd</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">vdata</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,<span class="va">dd</span><span class="op">)</span></span>
<span><span class="va">Vs</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/vst.html">vst</a></span><span class="op">(</span><span class="va">dd</span>,pops<span class="op">=</span><span class="va">pops</span><span class="op">)</span></span></code></pre></div>
<p><img src="rCNV_files/figure-html/unnamed-chunk-24-1.png" width="700">
Fig. 3.1. Vst network qgraph generated from the Vst values for each
lobster population. Circles represent populations and the green arrows
show similarity and the direction.See <code>qgraph</code> help for the
interpretation of the network analysis.</p>
<p>NOTE: You can also run a permutation test (using
<code>vstPermutation</code>) on the vst calculation to test the
randomness of the vst.</p>
</div>
<div class="section level4">
<h4 id="relative-copy-number">3.2 Relative copy number<a class="anchor" aria-label="anchor" href="#relative-copy-number"></a>
</h4>
<p>TBA</p>
</div>
</div>
<div class="section level3">
<h3 id="references">4. References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<ul>
<li><p>Chen, J., Li, L., Milesi, P., Jansson, G., Berlin, M., Karlsson,
B., … Lascoux, M. (2019). Genomic data provide new insights on the
demographic history and the extent of recent material transfers in
Norway spruce. Evolutionary Applications, 12(8), 1539–1551. doi:
10.1111/eva.12801</p></li>
<li><p>Dorant, Y., Cayuela, H., Wellband, K., Laporte, M., Rougemont,
Q., Mérot, C., … Bernatchez, L. (2020). Copy number variants outperform
SNPs to reveal genotype–temperature association in a marine species.
Molecular Ecology, 29(24), 4765–4782. doi: 10.1111/mec.15565</p></li>
<li><p>Karunarathne P, Zhou Q, Schliep K, Milesi P. A comprehensive
framework for detecting copy number variants from single nucleotide
polymorphism data: ‘rCNV’, a versatile r package for paralogue and CNV
detection. Mol Ecol Resour. 2023 Jul 29. doi:
10.1111/1755-0998.13843</p></li>
<li><p>Karunarathne, P., Zhou, Q., Schliep, K., &amp; Milesi, P. (2022).
A new framework for detecting copy number variants from single
nucleotide polymorphism data: ‘rCNV’, a versatile R package for paralogs
and CNVs detection. BioRxiv, 2022.10.14.512217. doi:
10.1101/2022.10.14.512217</p></li>
<li><p>Redon, R., Ishikawa, S., Fitch, K. R., Feuk, L., Perry, G. H.,
Andrews, T. D., … Chen, W. (2006). Global variation in copy number in
the human genome. Nature, 444(7118), 444–454.</p></li>
<li><p>Stockwell, B. L., Larson, W. A., Waples, R. K., Abesamis, R. A.,
Seeb, L. W., &amp; Carpenter, K. E. (2016). The application of genomics
to inform conservation of a functionally important reef fish (Scarus
niger) in the Philippines. Conservation Genetics, 17(1),
239–249</p></li>
<li><p>Yang, J., Benyamin, B., McEvoy, B. P., Gordon, S., Henders, A.
K., Nyholt, D. R., … Visscher, P. M. (2010). Common SNPs explain a large
proportion of the heritability for human height. Nature Genetics, 42(7),
565–569. doi: 10.1038/ng.608</p></li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Piyal Karunarathne, Qiujie Zhou, Klaus Schliep, Pascal Milesi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
