<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rCNV">
<title>Detecting CNV from SNPs: a quick guide to rCNV • rCNV</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Detecting CNV from SNPs: a quick guide to rCNV">
<meta property="og:description" content="rCNV">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rCNV</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/rCNV.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/piyalkarum/rCNV/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">
-
-  -

<div class="row">
  <main id="main"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Detecting CNV from SNPs: a quick guide to rCNV</h1>
                        <h4 data-toc-skip class="author">Piyal Karunarathne</h4>
            
            <h4 data-toc-skip class="date">Last updated: 04 May, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/piyalkarum/rCNV/blob/HEAD/vignettes/rCNV.Rmd" class="external-link"><code>vignettes/rCNV.Rmd</code></a></small>
      <div class="d-none name"><code>rCNV.Rmd</code></div>
    </div>

    
    
<div class="section level1">
<h1 id="rcnv-an-r-package-for-detecting-copy-number-variants-from-snps-data">
<strong>rCNV: An R package for detecting copy number variants from SNPs data</strong><a class="anchor" aria-label="anchor" href="#rcnv-an-r-package-for-detecting-copy-number-variants-from-snps-data"></a>
</h1>
<p><span style="color: dodgerblue;">Piyal Karunarathne, Qiujie Zhou, and Pascal Milesi</span></p>
<p>This tutorial provides a complete guide to detecting CNVs from SNPs data using rCNV package. This includes importing raw VCF files, filtering VCFs, normalization of read depth, categorization of SNPs into “deviants” and “non-deviants” from which “duplicates” and “singlets” are then recognized. A brief workflow is presented in Fig 1. below.</p>
<div class="figure">
<img src="workflow1.png" width="400" alt=""><p class="caption">Fig. 1. Workflow of detecting CNVs with rCNV</p>
</div>
<p>All the data used in this tutorial is available either withing the package or from the GitHub repository at <strong>link</strong>.</p>
<p>We provide here a complete workflow of detecting CNVs from SNPs, starting with raw (unfiltered) VCF as in the number 1 of the workflow chart. However, if one believes that they have filtered VCFs for the parameters we’ve mentioned (see Fig.1 and sub-sections 1.2 and 1.3), they can start from the number 2 in the workflow shown above.</p>
<div class="section level2">
<h2 id="pre-detection">1. PRE-DETECTION<a class="anchor" aria-label="anchor" href="#pre-detection"></a>
</h2>
<div class="section level3">
<h3 id="importing-data">1.1 Importing data<a class="anchor" aria-label="anchor" href="#importing-data"></a>
</h3>
<div class="line-block">    Raw VCF files can be imported to rCNV R environment using the built-in function <code><a href="../reference/readVCF.html">readVCF()</a></code>. The imported vcf is stored as a data frame within a list.</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Start by installing the package if you haven't already done so.  </span>
<span class="co"># To install the CRAN version</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">rCNV</span><span class="op">)</span>

<span class="co"># Or if you would like to install the development version of the package</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> 
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span> 
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"piyalkarum/rCNV"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://piyalkarum.github.io/rCNV/" class="external-link">rCNV</a></span><span class="op">)</span>
<span class="va">vcf.file.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"rCNV"</span><span class="op">)</span>, <span class="st">"/example.raw.vcf.gz"</span><span class="op">)</span>
<span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readVCF.html">readVCF</a></span><span class="op">(</span><span class="va">vcf.file.path</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># This imports an example VCF file included in the package</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#print the first 10 rows and columns of the imported VCF</span>
<span class="va">vcf</span><span class="op">$</span><span class="va">vcf</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">#&gt;    #CHROM   POS ID REF ALT      QUAL FILTER                 FORMAT</span>
<span class="co">#&gt; 1:  MA_10 24959  .   A   C   4338.29   PASS GT:AD:DP:GQ:PGT:PID:PL</span>
<span class="co">#&gt; 2:  MA_10 26811  .   C   T  33278.10   PASS         GT:AD:DP:GQ:PL</span>
<span class="co">#&gt; 3:  MA_10 26860  .   A   G 273719.00   PASS GT:AD:DP:GQ:PGT:PID:PL</span>
<span class="co">#&gt; 4:  MA_10 26874  .   T   C 374526.00   PASS GT:AD:DP:GQ:PGT:PID:PL</span>
<span class="co">#&gt; 5:  MA_31 21911  .   C   G   2069.06   PASS         GT:AD:DP:GQ:PL</span>
<span class="co">#&gt; 6:  MA_31 40129  .   A   T   6549.58   PASS         GT:AD:DP:GQ:PL</span>
<span class="co">#&gt;                              AT_PA_06_12</span>
<span class="co">#&gt; 1:             0/0:7,0:7:21:.:.:0,21,269</span>
<span class="co">#&gt; 2:               0/0:18,0:18:22:0,22,576</span>
<span class="co">#&gt; 3: 0/1:9,5:14:99:0|1:26860_A_G:183,0,775</span>
<span class="co">#&gt; 4:             0/0:15,0:15:3:.:.:0,3,407</span>
<span class="co">#&gt; 5:               0/0:11,0:11:33:0,33,319</span>
<span class="co">#&gt; 6:               0/0:11,0:11:33:0,33,347</span></code></pre>
</div>
<div class="section level3">
<h3 id="filtering-vcf-for-missing-data">1.2 Filtering VCF for missing data<a class="anchor" aria-label="anchor" href="#filtering-vcf-for-missing-data"></a>
</h3>
<div class="line-block">      The raw VCF files often contain missing data for both samples and SNPs. If retained, these samples and SNPs can introduce false positives and negatives to the analysis. Therefore, they must be removed based on a missing percentage. The function <code><a href="../reference/get.miss.html">get.miss()</a></code> determines the missing percentages in samples and SNPs and graphically visualize the missing ranges with 5% quantile values. We recommend removing samples and SNPs with missing percentages higher than 50%.</div>
<p>The following example demonstrates such filtering on Parrot Fish data from <em>McKinney et al. (2017)</em>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#download the parrotfish data file from the following link to your working directory if you do not have the github folder</span>
<span class="co">#https://drive.google.com/uc?export=download&amp;id=141-RRGKJaioGWuJxBOAz3D-nUmNI36LZ</span>
<span class="va">fl</span><span class="op">&lt;-</span><span class="st">"../testDATA/Parrotfish_sequenceReads.vcf.gz"</span>
<span class="va">parrot</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/readVCF.html">readVCF</a></span><span class="op">(</span><span class="va">fl</span>,verbose<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="va">mss</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/get.miss.html">get.miss</a></span><span class="op">(</span><span class="va">parrot</span>,verbose <span class="op">=</span> <span class="cn">F</span>, plot<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>This generates a list with missing percentage tables for SNPs and samples and a plot similar to Fig 1.2 where missing percentages are plotted. Users can also manually plot these percentages on their discretion. Further, if desired one type of missing percentage (i.e. per sample or per SNP) can be calculated separately by specifying the argument <code>type=</code> in the <code><a href="../reference/get.miss.html">get.miss()</a></code> function. See the help page of <code><a href="../reference/get.miss.html">get.miss()</a></code> for more information.</p>
<div class="figure">
<img src="get.miss.jpeg" style="width:80.0%" alt=""><p class="caption">Fig.1.2. Missing status of the VCF</p>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mss</span><span class="op">$</span><span class="va">perSample</span><span class="op">)</span>
<span class="co">#&gt;   indiv n_miss     f_miss</span>
<span class="co">#&gt; 1  DP_1   1089 0.05726455</span>
<span class="co">#&gt; 2 DP_10   1086 0.05710680</span>
<span class="co">#&gt; 3 DP_11   1365 0.07177788</span>
<span class="co">#&gt; 4 DP_12   1091 0.05736972</span>
<span class="co">#&gt; 5 DP_13   1069 0.05621286</span>
<span class="co">#&gt; 6 DP_14    387 0.02035021</span></code></pre></div>
<p>The plot shows that there are several samples with missing data for more than 50% of the SNPs. It is recommended to remove these before proceeding with further analysis. Removing them from the imported vcf file is just a matter of obtaining the columns (samples) for which missing percentage if &gt;50%.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">$</span><span class="va">vcf</span><span class="op">)</span>
<span class="co">#&gt; [1] 19017   105</span>
<span class="va">sam</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">mss</span><span class="op">$</span><span class="va">perSample</span><span class="op">$</span><span class="va">f_miss</span><span class="op">&gt;</span><span class="fl">0.5</span><span class="op">)</span><span class="op">+</span><span class="fl">9</span> <span class="co">#here we determine the sample columns with f_miss&gt;0.5 and add 9 to get the correct column number</span>
<span class="va">parrot</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">$</span><span class="va">vcf</span><span class="op">)</span><span class="op">[</span>,<span class="op">-</span><span class="va">sam</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">)</span>
<span class="co">#&gt; [1] 19017   100</span></code></pre></div>
<p>We have removed 5 samples that were undesired for the analysis due to their missing data. Undesired SNPs can also be removed the same way; Hint: now they are in rows.</p>
</div>
<div class="section level3">
<h3 id="assessing-the-relatedness-among-samples-and-heterozygosity-withing">1.3 Assessing the relatedness among samples and heterozygosity withing<a class="anchor" aria-label="anchor" href="#assessing-the-relatedness-among-samples-and-heterozygosity-withing"></a>
</h3>
<p>The function <code><a href="../reference/relatedness.html">relatedness()</a></code> in the rCNV package will generate the relatedness index between pairs of samples according to <em>Yang et al. (2010)</em>. The output is a data frame with pairs of samples and the relatedness score <em>Ajk</em>. Please read <em>Yang et al. (2010)</em> for more information on the calculation of <em>Ajk</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rels</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/relatedness.html">relatedness</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rels</span><span class="op">)</span></code></pre></div>
<p>From the relatedness plots, we can see that the relatedness in Parrot fish samples is low. We recommend removing one sample of the pairs with outlier values above 0.9.</p>
<p><img src="relatedness.png" style="width:40.0%"><img src="h.zygosity.png" style="width:40.0%"></p>
<p>Figure 1.3.a Relatedness among samples,b Heterozygosity in Parrotfish populations</p>
<p>Heterozygosity within populations can be calclated using <code><a href="../reference/h.zygosity.html">h.zygosity()</a></code> function.</p>
<div class="line-block">  * NOTE: A list of populations corresponding to the samples must be provided in order to plot the heterozygosity among populations.</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pops</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">parrot</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span> <span class="co">#population codes (the first two letters of the sample names)</span>
<span class="va">hz</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/h.zygosity.html">h.zygosity</a></span><span class="op">(</span><span class="va">parrot</span>,pops<span class="op">=</span><span class="va">pops</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hz</span><span class="op">)</span></code></pre></div>
<p>In figure 1.3.b, we can see that most of the individuals has a mean heterozygosity F<sub>IS</sub> close to 0.19, except for a few outliers. However, outliers with an F<sub>IS</sub> value below -0.2 is highly likely DNA contamination. Therefore, it is recommended to remove such samples.</p>
</div>
</div>
<div class="section level2">
<h2 id="generating-allele-depth-tables-and-normalized-depth-values">1.4 Generating allele depth tables and normalized depth values<a class="anchor" aria-label="anchor" href="#generating-allele-depth-tables-and-normalized-depth-values"></a>
</h2>
<p>Allele read depth table is the main data set that will be used for the rest of the analysis. Therefore, having them generated and stored for later use will immensely save a lot of time. The function <code><a href="../reference/hetTgen.html">hetTgen()</a></code> is especially dedicated for this purpose. Apart from generating filtered allele depth table for separate alleles (<code>info.type="AD"</code>), the function can also generate total allele depth (<code>info.type="AD-tot"</code>), genotype table (<code>info.type="GT"</code>), etc.. See the help page of <code><a href="../reference/hetTgen.html">hetTgen()</a></code> for more formats.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ad.tab</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/hetTgen.html">hetTgen</a></span><span class="op">(</span><span class="va">parrot</span>,info.type<span class="op">=</span><span class="st">"AD"</span><span class="op">)</span>
<span class="va">ad.tab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></code></pre></div>
Table 1.4.0 Allele depth table (generated from a different data set than Parrot fish data)
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
CHROM
</th>
<th style="text-align:right;">
POS
</th>
<th style="text-align:right;">
ID
</th>
<th style="text-align:left;">
ALT
</th>
<th style="text-align:left;">
KKOGR07_0005_combo
</th>
<th style="text-align:left;">
KKOGR07_0006
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
7452
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
705339
</td>
<td style="text-align:right;">
67808
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
45,0
</td>
<td style="text-align:left;">
34,0
</td>
</tr>
<tr>
<td style="text-align:left;">
18079
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
769896
</td>
<td style="text-align:right;">
80692
</td>
<td style="text-align:left;">
T
</td>
<td style="text-align:left;">
16,0
</td>
<td style="text-align:left;">
15,0
</td>
</tr>
<tr>
<td style="text-align:left;">
9793
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
926883
</td>
<td style="text-align:right;">
109573
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
56,0
</td>
<td style="text-align:left;">
42,0
</td>
</tr>
<tr>
<td style="text-align:left;">
11755
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
172395
</td>
<td style="text-align:right;">
16363
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
61,67
</td>
<td style="text-align:left;">
40,34
</td>
</tr>
<tr>
<td style="text-align:left;">
10122
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
18102
</td>
<td style="text-align:right;">
1703
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
75,0
</td>
<td style="text-align:left;">
51,0
</td>
</tr>
<tr>
<td style="text-align:left;">
816
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
76725
</td>
<td style="text-align:right;">
6991
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
66,41
</td>
<td style="text-align:left;">
20,20
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#normalize depth table with cpm.normal()</span>
<span class="va">ad.nor</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/cpm.normal.html">cpm.normal</a></span><span class="op">(</span><span class="va">ad.tab</span><span class="op">)</span>
<span class="va">ad.nor</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></code></pre></div>
<p>Table 1.4.1 Normalized allele depth table</p>
<pre><code><span class="co">#&gt; OUTLIERS DETECTED</span>
<span class="co">#&gt; Consider removing the samples:</span>
<span class="co">#&gt; KKOGR07_0030 KKOGR07_0043_combo KKOKT10_0069</span></code></pre>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
CHROM
</th>
<th style="text-align:right;">
POS
</th>
<th style="text-align:right;">
ID
</th>
<th style="text-align:left;">
ALT
</th>
<th style="text-align:left;">
KKOGR07_0005_combo
</th>
<th style="text-align:left;">
KKOGR07_0006
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
7452
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
705339
</td>
<td style="text-align:right;">
67808
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
228.93,0
</td>
<td style="text-align:left;">
268.65,0
</td>
</tr>
<tr>
<td style="text-align:left;">
18079
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
769896
</td>
<td style="text-align:right;">
80692
</td>
<td style="text-align:left;">
T
</td>
<td style="text-align:left;">
81.4,0
</td>
<td style="text-align:left;">
118.52,0
</td>
</tr>
<tr>
<td style="text-align:left;">
9793
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
926883
</td>
<td style="text-align:right;">
109573
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
284.89,0
</td>
<td style="text-align:left;">
331.87,0
</td>
</tr>
<tr>
<td style="text-align:left;">
11755
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
172395
</td>
<td style="text-align:right;">
16363
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
310.32,340.85
</td>
<td style="text-align:left;">
316.06,268.65
</td>
</tr>
<tr>
<td style="text-align:left;">
10122
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
18102
</td>
<td style="text-align:right;">
1703
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
381.55,0
</td>
<td style="text-align:left;">
402.98,0
</td>
</tr>
<tr>
<td style="text-align:left;">
816
</td>
<td style="text-align:left;">
un
</td>
<td style="text-align:right;">
76725
</td>
<td style="text-align:right;">
6991
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
335.76,208.58
</td>
<td style="text-align:left;">
158.03,158.03
</td>
</tr>
</tbody>
</table>
<p>Compare the allele depth values between tables 1.4.0 and 1.4.1.</p>
</div>
<div class="section level2">
<h2 id="detection">2. DETECTION<a class="anchor" aria-label="anchor" href="#detection"></a>
</h2>
<div class="section level3">
<h3 id="generating-allele-information-table">2.1 Generating allele information table<a class="anchor" aria-label="anchor" href="#generating-allele-information-table"></a>
</h3>
<p>In this step, all the statistics necessary for detecting duplicates are calculated. They include the following: a. Allele ratios (across all samples), b. Proportions of homo-/heterozygotes per SNP, c. Depth ratios (alternative/sum), d. Z-score, e. Chi-square significance, and f. excess of heterozygosity</p>
<p>Alternative allele ratio is calculated across all the samples from depth values;Proportion of homo/hetero-zygotes are also calculated all the samples per SNP; Depth ratio is calculated per individual per SNP by dividing the alternative allele depth value by the sum depth value of both alleles. This is calculated for both heterozygotes and homozygotes; Z-score per SNP is calculated according to the following equation:</p>
<p><img src="eq.svg"></p>
<p><img src="eq2.svg"></p>
<p>Where N is the total depth summed for all heterozygotes, N<sub>A</sub> is the depth of allele A summed for all the heterozygotes, <em>P</em> is the occurrence probability of allele A in heterozygotes - for unbiased sequencing this is 0.5. <code><a href="../reference/allele.info.html">allele.info()</a></code> function calculates this for both 0.5 and biased probability using the ratio between reference and alternative alleles.</p>
<p>The Chi-squared values per SNP per sample were calculated using the following equation: <img src="eq.chi.svg"></p>
<p>Where <em>N<sub>jk</sub></em> is the total depth value of SNP <em>k</em> for individual <em>j</em>, <em>D<sub>jk</sub></em> is the depth value of the alternative allele of SNP <em>k</em>, <em>P<sub>j</sub></em> is the occurrence probability of alternative allele at SNP <em>k</em> in heterozygotes - in unbiased sequencing, this is 0.5.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">A.info</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/allele.info.html">allele.info</a></span><span class="op">(</span><span class="va">ADtable</span>, plot.allele.cov <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">A.info</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt;       CHROM    POS     ID NHet    propHet  medRatio NHomRef NHomAlt propHomAlt</span>
<span class="co">#&gt; 7452     un 705339  67808    9 0.08653846 0.5000000      95       0 0.00000000</span>
<span class="co">#&gt; 18079    un 769896  80692   41 0.40594059 0.5000000      48      12 0.11881188</span>
<span class="co">#&gt; 9793     un 926883 109573   54 0.51923077 0.5096539      34      16 0.15384615</span>
<span class="co">#&gt; 10122    un  18102   1703    8 0.07619048 0.5174242      97       0 0.00000000</span>
<span class="co">#&gt; 816      un  76725   6991   49 0.47115385 0.4897959      35      20 0.19230769</span>
<span class="co">#&gt; 17482    un 713297  72437   46 0.43809524 0.5217593      52       7 0.06666667</span>
<span class="co">#&gt;       Nsamp      pAll      pHet       fis     z.het      z.05     z.all</span>
<span class="co">#&gt; 7452    104 0.5516892 0.5155351 0.9899497 0.9853781 0.5373322 0.1669619</span>
<span class="co">#&gt; 18079   101 0.5082649 0.4825144 0.9907917 0.9531631 0.3160515 0.1473500</span>
<span class="co">#&gt; 9793    104 0.5043312 0.4939215 0.9897064 0.6534022 0.8933630 0.5826420</span>
<span class="co">#&gt; 10122   105 0.4615933 0.4916555 0.9900990 0.7967226 0.5758174 0.4050012</span>
<span class="co">#&gt; 816     104 0.5007026 0.5002302 0.9907469 0.6409348 0.6562743 0.6099265</span>
<span class="co">#&gt; 17482   105 0.5086655 0.5166121 0.9897778 0.9829558 0.1245431 0.4556636</span>
<span class="co">#&gt;         chi.het    chi.05   chi.all   z.het.sum    z.05.sum z.all.sum</span>
<span class="co">#&gt; 7452  0.6129256 0.5563662 0.5912685 -0.05498053  -1.8505568  4.146095</span>
<span class="co">#&gt; 18079 0.8685639 0.8845710 0.8829935  0.37608802   6.4198038  9.277843</span>
<span class="co">#&gt; 9793  0.8196782 0.8368758 0.8426532 -3.29976321   0.9850626  4.038202</span>
<span class="co">#&gt; 10122 0.2941631 0.3026659 0.2062024  0.72857723   1.5825180 -2.355295</span>
<span class="co">#&gt; 816   0.6488278 0.6486573 0.6491177  3.26474533   3.1154399  3.571248</span>
<span class="co">#&gt; 17482 0.7487560 0.6250022 0.7040393 -0.14489338 -10.4175272 -5.059661</span>
<span class="co">#&gt;       chi.het.sum chi.0.5.sum chi.all.sum   eH.pval   eH.delta</span>
<span class="co">#&gt; 7452     6.306639    6.818223    6.501254 1.0000000  0.3894231</span>
<span class="co">#&gt; 18079   30.235077   29.656595   29.715671 0.8441666 -3.0841584</span>
<span class="co">#&gt; 9793    43.540437   42.933241   42.721388 0.8797139  3.5576923</span>
<span class="co">#&gt; 10122    8.455278    8.350960    9.700308 1.0000000  0.3047619</span>
<span class="co">#&gt; 816     43.717658   43.721894   43.710452 0.9656811 -1.9182692</span>
<span class="co">#&gt; 17482   38.324345   41.405597   39.483216 0.8303502  3.1428571</span></code></pre>
<p>In addition to the above a-e parameters several other important statistics per sample per SNP is generated in the <code><a href="../reference/allele.info.html">allele.info()</a></code> output. See below. ADD MORE ABOUT THE PROBE BIASE PLOTS****</p>
</div>
<div class="section level3">
<h3 id="deviants-detection">2.2 Deviants detection<a class="anchor" aria-label="anchor" href="#deviants-detection"></a>
</h3>
<p>In this step, we use the allele information table generated in the previous step to detect deviant SNPs based different models, which then used in the next step with the appropriate test statistic to detect duplicates. The rCNV function <code><a href="../reference/dupGet.html">dupGet()</a></code> delivers easy to pick model selections to detect deviant SNPs with a plotting (similar to HD plots in <em>McKinney et al. 2017</em>) option to visualize the detection.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run this code for a demonstration of the detection</span>
<span class="va">dvs</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/dupGet.html">dupGet</a></span><span class="op">(</span><span class="va">alleleINF</span><span class="op">)</span></code></pre></div>
<p>The detection of deviant SNPs is based on the <strong>heterozygosity</strong> within populations (or among all samples) and <strong>allelic ratios</strong> within such heterozygotes. As such, <code><a href="../reference/dupGet.html">dupGet()</a></code> function implements three methods to flag the deviant alleles; They are a) Excess of heterozygotes and b) Heterozygote SNPs that do not follow normal Z distribution, and c) Heterozygote SNPs that do not follow normal chi-square distribution.</p>
<div class="line-block">  The proportion of heterozygotes in a population according to HWE is 2pq where as homozygotes is q<sup>2</sup>. In deviant SNPs heterozygozity is expected to be higher indicating excess of heterozygosity and thus flagged as deviant/putative-duplicates.</div>
<div class="line-block">  In non-duplicated alleles (singlets), median allele ratio must be ~0.5, however, in deviants it varies between 1:3 and 3:1. These SNP deviants are also flagged as putative duplicates. To determine the deviation, we use the Z-score values and Chi-square significance calculated in the previous step.</div>
<div class="line-block">  **NOTE: users can select the occurrence probability of the alternative allele based on the sequencing method and the probe bias therein.</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">deviants</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/dupGet.html">dupGet</a></span><span class="op">(</span><span class="va">alleleINF</span>,test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"z.all"</span>,<span class="st">"chi.all"</span><span class="op">)</span>,plot<span class="op">=</span><span class="cn">TRUE</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deviants</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt;       CHROM    POS     ID NHet    propHet  medRatio NHomRef NHomAlt propHomAlt</span>
<span class="co">#&gt; 7452     un 705339  67808    9 0.08653846 0.5000000      95       0 0.00000000</span>
<span class="co">#&gt; 18079    un 769896  80692   41 0.40594059 0.5000000      48      12 0.11881188</span>
<span class="co">#&gt; 9793     un 926883 109573   54 0.51923077 0.5096539      34      16 0.15384615</span>
<span class="co">#&gt; 10122    un  18102   1703    8 0.07619048 0.5174242      97       0 0.00000000</span>
<span class="co">#&gt; 816      un  76725   6991   49 0.47115385 0.4897959      35      20 0.19230769</span>
<span class="co">#&gt; 17482    un 713297  72437   46 0.43809524 0.5217593      52       7 0.06666667</span>
<span class="co">#&gt;       Nsamp   eH.pval   eH.delta    dup.stat</span>
<span class="co">#&gt; 7452    104 1.0000000  0.3894231 non-deviant</span>
<span class="co">#&gt; 18079   101 0.8441666 -3.0841584 non-deviant</span>
<span class="co">#&gt; 9793    104 0.8797139  3.5576923 non-deviant</span>
<span class="co">#&gt; 10122   105 1.0000000  0.3047619 non-deviant</span>
<span class="co">#&gt; 816     104 0.9656811 -1.9182692 non-deviant</span>
<span class="co">#&gt; 17482   105 0.8303502  3.1428571 non-deviant</span></code></pre>
<p>The function also plots the deviants with Allele median ratio Vs Proportion of Heterozygotes to visualize and validate the detection by the user. Users can also plot the detection separately with the output of <code><a href="../reference/dupGet.html">dupGet()</a></code> using <code><a href="../reference/dup.plot.html">dup.plot()</a></code> fucntion.</p>
<div class="figure">
<img src="dupPlot.png" width="400" alt=""><p class="caption">Fig 2.1 Median allele ratio Vs Prop. of Heterozygotes with putative duplicates</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="post-detection">3. POST-DETECTION<a class="anchor" aria-label="anchor" href="#post-detection"></a>
</h2>
<div class="line-block">      TBA</div>
<div class="section level3">
<h3 id="duplicate-list-and-validation">3.1 Duplicate list and validation<a class="anchor" aria-label="anchor" href="#duplicate-list-and-validation"></a>
</h3>
</div>
<div class="section level3">
<h3 id="relative-copy-number">3.2 Relative copy number<a class="anchor" aria-label="anchor" href="#relative-copy-number"></a>
</h3>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Piyal Karunarathne, Qiujie Zhou, Pascal Milesi.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
