<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get allele information for duplicate detection — allele.info • rCNV</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get allele information for duplicate detection — allele.info"><meta name="description" content="The function to calculate allele median ratios, proportion of heterozygotes
and allele probability values under different assumptions (see details),
and their chi-square significance values for duplicate detection"><meta property="og:description" content="The function to calculate allele median ratios, proportion of heterozygotes
and allele probability values under different assumptions (see details),
and their chi-square significance values for duplicate detection"><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rCNV</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rCNV.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/piyalkarum/rCNV/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Get allele information for duplicate detection</h1>
      <small class="dont-index">Source: <a href="https://github.com/piyalkarum/rCNV/blob/HEAD/R/bias.det.R" class="external-link"><code>R/bias.det.R</code></a></small>
      <div class="d-none name"><code>allele.info.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function to calculate allele median ratios, proportion of heterozygotes
and allele probability values under different assumptions (see details),
and their chi-square significance values for duplicate detection</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">allele.info</span><span class="op">(</span></span>
<span>  <span class="va">X</span>,</span>
<span>  x.norm <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">Fis</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MedR"</span>, <span class="st">"QN"</span>, <span class="st">"pca"</span>, <span class="st">"TMM"</span>, <span class="st">"TMMex"</span><span class="op">)</span>,</span>
<span>  logratioTrim <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  sumTrim <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  Weighting <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Acutoff <span class="op">=</span> <span class="op">-</span><span class="fl">1e+10</span>,</span>
<span>  plot.allele.cov <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">X<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>allele depth table generated from the function
<code>hetTgen</code> (non-normalized)</p></dd>


<dt id="arg-x-norm">x.norm<a class="anchor" aria-label="anchor" href="#arg-x-norm"></a></dt>
<dd><p>a data frame of normalized allele coverage, output of
<code>cpm.normal</code>. If not provided, calculated using <code>X</code>.</p></dd>


<dt id="arg-fis">Fis<a class="anchor" aria-label="anchor" href="#arg-fis"></a></dt>
<dd><p>numeric. Inbreeding coefficient calculated using <code><a href="h.zygosity.html">h.zygosity()</a></code> function</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>character. method to be used for normalization
(see <code>cpm.normal</code> details). Default <code>TMM</code></p></dd>


<dt id="arg-logratiotrim">logratioTrim<a class="anchor" aria-label="anchor" href="#arg-logratiotrim"></a></dt>
<dd><p>numeric. percentage value (0 - 1) of variation to be
trimmed in log transformation</p></dd>


<dt id="arg-sumtrim">sumTrim<a class="anchor" aria-label="anchor" href="#arg-sumtrim"></a></dt>
<dd><p>numeric. amount of trim to use on the combined absolute
levels (“A” values) for method <code>TMM</code></p></dd>


<dt id="arg-weighting">Weighting<a class="anchor" aria-label="anchor" href="#arg-weighting"></a></dt>
<dd><p>logical, whether to compute (asymptotic binomial precision)
weights</p></dd>


<dt id="arg-acutoff">Acutoff<a class="anchor" aria-label="anchor" href="#arg-acutoff"></a></dt>
<dd><p>numeric, cutoff on “A” values to use before trimming</p></dd>


<dt id="arg-plot-allele-cov">plot.allele.cov<a class="anchor" aria-label="anchor" href="#arg-plot-allele-cov"></a></dt>
<dd><p>logical, plot comparative plots of allele depth
coverage in homozygotes and heterozygotes</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>logical, whether to print progress</p></dd>


<dt id="arg-parallel">parallel<a class="anchor" aria-label="anchor" href="#arg-parallel"></a></dt>
<dd><p>logical. whether to parallelize the process</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>further arguments to be passed to <code>plot</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a data frame of median allele ratio, proportion of
heterozygotes, number of heterozygotes, and allele probability at different
assumptions with their chi-square significance</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Allele information generated here are individual SNP based and presents the
proportion of heterozygotes, number of samples, and deviation of allele
detection from a 1:1 ratio of reference and alternative alleles.
The significance of the deviation is tested with Z-score test
\(Z = \frac{ \frac{N}{2}-N_A}{ \sigma_{x}}\),
and chi-square test (see references for more details on the method).</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>McKinney, G. J., Waples, R. K., Seeb, L. W., &amp; Seeb, J. E. (2017).
Paralogs are revealed by proportion of heterozygotes and deviations in read
ratios in genotyping by sequencing data from natural populations.
Molecular Ecology Resources, 17(4)</p></li>
<li><p>Karunarathne et al. 2022 (to be added)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Piyal Karunarathne, Pascal Milesi, Klaus Schliep</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">hz</span><span class="op">&lt;-</span><span class="fu"><a href="h.zygosity.html">h.zygosity</a></span><span class="op">(</span><span class="va">vcf</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">Fis</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">hz</span><span class="op">$</span><span class="va">Fis</span>,na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ADtable</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">AI</span><span class="op">&lt;-</span><span class="fu">allele.info</span><span class="op">(</span><span class="va">ADtable</span>,x.norm<span class="op">=</span><span class="va">ADnorm</span>,Fis<span class="op">=</span><span class="fl">0.11</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Piyal Karunarathne, Qiujie Zhou, Klaus Schliep, Pascal Milesi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

