<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate normalization factor for each sample — norm.fact • rCNV</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate normalization factor for each sample — norm.fact"><meta name="description" content="This function calculates the normalization factor for each sample using
different methods. See details."><meta property="og:description" content="This function calculates the normalization factor for each sample using
different methods. See details."><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rCNV</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.900</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rCNV.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/piyalkarum/rCNV/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculate normalization factor for each sample</h1>
      <small class="dont-index">Source: <a href="https://github.com/piyalkarum/rCNV/blob/HEAD/R/TMM_normalz.R" class="external-link"><code>R/TMM_normalz.R</code></a></small>
      <div class="d-none name"><code>norm.fact.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates the normalization factor for each sample using
different methods. See details.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">norm.fact</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TMM"</span>, <span class="st">"TMMex"</span>, <span class="st">"MedR"</span>, <span class="st">"QN"</span><span class="op">)</span>,</span>
<span>  logratioTrim <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  sumTrim <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  Weighting <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Acutoff <span class="op">=</span> <span class="op">-</span><span class="fl">1e+10</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-df">df<a class="anchor" aria-label="anchor" href="#arg-df"></a></dt>
<dd><p>a data frame or matrix of allele depth values
(total depth per snp per sample)</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>character. method to be used (see details). Default <code>TMM</code></p></dd>


<dt id="arg-logratiotrim">logratioTrim<a class="anchor" aria-label="anchor" href="#arg-logratiotrim"></a></dt>
<dd><p>numeric. percentage value (0 - 1) of variation to be
trimmed in log transformation</p></dd>


<dt id="arg-sumtrim">sumTrim<a class="anchor" aria-label="anchor" href="#arg-sumtrim"></a></dt>
<dd><p>numeric. amount of trim to use on the combined absolute
levels (“A” values) for method <code>TMM</code></p></dd>


<dt id="arg-weighting">Weighting<a class="anchor" aria-label="anchor" href="#arg-weighting"></a></dt>
<dd><p>logical, whether to compute (asymptotic binomial precision)
weights</p></dd>


<dt id="arg-acutoff">Acutoff<a class="anchor" aria-label="anchor" href="#arg-acutoff"></a></dt>
<dd><p>numeric, cutoff on “A” values to use before trimming</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a numerical vector of normalization factors for each sample</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Originally described for normalization of RNA sequences
(Robinson &amp; Oshlack 2010), this function computes normalization (scaling)
factors to convert observed library sizes into effective library sizes.
It uses the method trimmed means of M-values proposed by Robinson &amp;
Oshlack (2010). See the original publication and <code>edgeR</code> package
for more information.
The method <code>MedR</code> is median ratio normalization;
QN - quantile normalization (see  Maza, Elie, et al. 2013 for a
comparison of methods).</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Robinson MD, and Oshlack A (2010). A scaling normalization method for
differential expression analysis of RNA-seq data. Genome Biology 11, R25</p></li>
<li><p>Robinson MD, McCarthy DJ and Smyth GK (2010). edgeR: a Bioconductor
package for differential expression analysis of digital gene expression
data. Bioinformatics 26</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Piyal Karunarathne</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="va">vcf.file.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"rCNV"</span><span class="op">)</span>, <span class="st">"/example.raw.vcf.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="readVCF.html">readVCF</a></span><span class="op">(</span><span class="va">vcf.file.path</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'vcf.file.path' not found</span>
<span class="r-in"><span><span class="va">df</span><span class="op">&lt;-</span><span class="fu"><a href="hetTgen.html">hetTgen</a></span><span class="op">(</span><span class="va">vcf</span>,<span class="st">"AD-tot"</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'vcf' not found</span>
<span class="r-in"><span><span class="fu">norm.fact</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="co"># \dontrun{}</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in if (method == "TMM") {    f75 &lt;- suppressWarnings(calcFactorQuantile(data = as.matrix(df),         lib.size = colSums(as.matrix(df)), p = 0.75))    if (median(f75) &lt; 1e-20) {        ref &lt;- which.max(colSums(sqrt(as.matrix(df))))    }    else {        ref &lt;- which.min(abs(f75 - mean(f75)))    }    out &lt;- apply(df, 2, FUN = TMM, ref = df[, ref], logratioTrim = logratioTrim,         sumTrim = sumTrim, Weighting = Weighting, Acutoff = Acutoff)} else if (method == "TMMex") {    ref &lt;- which.max(colSums(sqrt(as.matrix(df))))    out &lt;- apply(df, 2, FUN = TMMex, ref = df[, ref], logratioTrim = logratioTrim,         sumTrim = sumTrim, Weighting = Weighting, Acutoff = Acutoff)} else if (method == "MedR") {    pseudo &lt;- apply(df, 1, function(xx) {        exp(mean(log(as.numeric(xx)[as.numeric(xx) &gt; 0])))    })    out &lt;- apply(df, 2, function(xx) {        median(as.numeric(xx)/pseudo, na.rm = T)    })}:</span> the condition has length &gt; 1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Piyal Karunarathne, Qiujie Zhou, Klaus Schliep, Pascal Milesi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

